src/Symfony/Component/Console/Attribute/Input.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace Symfony\Component\Console\Attribute;
+
+use Symfony\Component\Console\Exception\LogicException;
+use Symfony\Component\Console\Input\InputInterface;
+
+#[\Attribute(\Attribute::TARGET_PARAMETER)]
+class Input
+{
+    private \ReflectionClass $class;
+    /**
+     * @var array<string, Argument>
+     */
+    private array $arguments = [];
+    /**
+     * @var array<string, Option>
+     */
+    private array $options = [];
+
+    public static function tryFrom(\ReflectionParameter $parameter): ?self
+    {
+        /** @var self $self */
+        if (!$self = ($parameter->getAttributes(self::class, \ReflectionAttribute::IS_INSTANCEOF)[0] ?? null)?->newInstance()) {
+            return null;
+        }
+
+        $type = $parameter->getType();
+
+        if (!$type instanceof \ReflectionNamedType) {
+            throw new LogicException(\sprintf('The input parameter "%s" must have a named type.', $parameter->name));
+        }
+
+        if (!class_exists($class = $type->getName())) {
+            throw new LogicException(\sprintf('The input class "%s" does not exist.', $type->getName()));
+        }
+
+        $self->class = new \ReflectionClass($class);
+
+        foreach ($self->class->getProperties() as $property) {
+            if ($argument = Argument::tryFrom($property)) {
+                $self->arguments[$property->name] = $argument;
+            } elseif ($option = Option::tryFrom($property)) {
+                $self->options[$property->name] = $option;
+            }
+        }
+
+        if ([] === $self->arguments && [] === $self->options) {
+            throw new LogicException(\sprintf('The input class "%s" must have at least one argument or option.', $self->class->name));
+        }
+
+        return $self;
+    }
+
+    /**
+     * @internal
+     */
+    public function resolveValue(InputInterface $input): mixed
+    {
+        $instance = $this->class->newInstanceWithoutConstructor();
+
+        foreach ($this->arguments as $name => $argument) {
+            $this->class->getProperty($name)->setValue($instance, $argument->resolveValue($input));
+        }
+
+        foreach ($this->options as $name => $option) {
+            $this->class->getProperty($name)->setValue($instance, $option->resolveValue($input));
```

Comment:
the current code bypasses visibility constraints, better not IMHO:
```suggestion
            $instance->$name = $option->resolveValue($input);
```

Reply:
I was hoping to define private properties that, once hydrated, are only accessible through getters. This allows better use of encapsulation, for example when splitting a raw value into multiple derived values without exposing it directly.

Reply:
We don't do that even in DI - where we could eg inject service in private properties, but we don't :)

Reply:
ah: readonly - there I'd be fine doing the injection, because honeslty, readonly is a mess so I've no concerns working around its quircks

Reply:
last thought: maybe map by constructor arg's names instead of property names? There we would have proper encapsulation for the hydratation

Reply:
yep constructor args is a good alternative for encapsulation, now I want both ;)

would it work for you if we only hydrate public non-promoted properties + any promoted ones?

Reply:
I kept it simple so far, allowing only public properties for now.